fgpyo/sam/__init__.py

```diff
@@ -1231,3 +1263,234 @@ class SamOrder(enum.Enum):
+
+        if len(stranded_start) <= 1 or stranded_start[0] not in ("+", "-"):
+            raise ValueError(f"The stranded start field is malformed: {stranded_start}")
+
+        return cls(
+            reference_name=reference_name,
+            reference_start=int(stranded_start[1:]) - 1,
+            is_forward=stranded_start[0] == "+",
+            cigar=Cigar.from_cigarstring(cigar),
+            edit_distance=int(edit_distance),
+            alignment_score=None if alignment_score is None else int(alignment_score),
+            mapq=None if mapq is None else int(mapq),
+        )
+
+    @classmethod
+    def many_from_tag(cls, value: str) -> list[Self]:
+        """Build many secondary alignments from a single `XA` or `XB` tag value.
+
+        Args:
+            value: A single `XA` or `XB` tag value.
+        """
+        return [cls.from_tag_part(part) for part in value.rstrip(";").split(";")]
+
+    @classmethod
+    def many_from_rec(cls, rec: AlignedSegment) -> list[Self]:
+        """Build many secondary alignments from a single SAM record.
+
+        Args:
+            rec: The SAM record to generate secondary alignments from.
+        """
+        secondaries: list[Self] = []
+        try:
+            if rec.has_tag("XA"):
+                secondaries.extend(cls.many_from_tag(cast(str, rec.get_tag("XA"))))
+            if rec.has_tag("XB"):
+                secondaries.extend(cls.many_from_tag(cast(str, rec.get_tag("XB"))))
+        except ValueError as exception:
+            raise ValueError(
+                f"Could not parse secondary alignments for read: {rec.query_name}"
+            ) from exception
+        return secondaries
+
+    @classmethod
+    def many_sam_from_rec(cls, rec: AlignedSegment) -> Iterator[AlignedSegment]:
+        """Build many SAM secondary alignments from a single SAM record.
+
+        All reconstituted secondary alignments will have the `rh` SAM tag set upon them.
+
+        By default, the query bases and qualities of the secondary alignment will be set to the
+        query bases and qualities of the record that created the secondary alignments. However, if
+        there are hard-clips in the record used to create the secondary alignments, then this
+        function will set the query qualities and bases to the space-saving and/or unknown marker
+        `*`. A future development for this function should correctly pad-out (with No-calls) or clip
+        the query sequence and qualities depending on the hard-clipping found in both ends of the
+        source (often a primary) record and both ends of the destination (secondary) record.
+
+        Args:
+            rec: The SAM record to generate secondary alignments from.
+        """
+        if (
+            rec.is_unmapped
+            or rec.cigarstring is None
+            or rec.query_sequence is None
+            or rec.query_qualities is None
+        ):
+            return
+
+        for hit in cls.many_from_rec(rec):
```

Comment:
do you want to move the body of this loop into its own method.  I.e. would it be useful to have a `to_pysam(self): AlignedSegment` method?